### PR TITLE
Refactor lowering

### PR DESCRIPTION
**Pull request by csarofeen**
Fully functional. Still some todo on cleaning. Only place left for cleanup is compute at maps. I want to narrow down precisely what's needed there and remove some unused items.

### CODE REVIEW

File: torch/csrc/jit/codegen/cuda/ir_iostream.cpp [hunk 1]
```diff
@@ -68,7 +68,8 @@ void IrPrinter::handle(const TensorView* tv) {
     if (tv->getComputeAtView() != nullptr) {
       os_ << " compute_at( ";
       os_ << "T" << tv->getComputeAtView()->name();
```
Comment (csarofeen) [inline torch/csrc/jit/codegen/cuda/ir_iostream.cpp]: Can be reverted once we change computeAt structure, but for now should leave in for debugging.

File: torch/csrc/jit/codegen/cuda/kernel_ir_printer.cpp [hunk 2]
```diff
@@ -1,6 +1,7 @@
 #include <torch/csrc/jit/codegen/cuda/kernel_ir_printer.h>
 
 #include <torch/csrc/jit/codegen/cuda/instrumentation.h>
+#include <torch/csrc/jit/codegen/cuda/ir_iostream.h>
```
Comment (csarofeen) [inline torch/csrc/jit/codegen/cuda/kernel_ir_printer.cpp]: changes in this file should be removed, or split into another PR, still needed for debugging.

File: torch/csrc/jit/codegen/cuda/lower_allocation.cpp [hunk 3]
```diff
@@ -113,10 +112,12 @@ class AllocationInserter : public kir::MutableIrVisitor {
...
       }
-      auto ca_id =
-          gpu_lower->lowerValue(fuser_tv->getComputeAtAxis(axis_i).first)
+      auto concrete_id =
+          gpu_lower
+              ->lowerValue(gpu_lower->caLoopMap().getConcreteMappedID(
+                  fuser_tv->axis(axis_i)))
               ->as<kir::IterDomain>();
-      init_dims.push_back(ca_id);
+      init_dims.push_back(concrete_id);
```
Comment (csarofeen) [inline torch/csrc/jit/codegen/cuda/lower_allocation.cpp:121]: @naoyam is this right? Should we allocate/init based on a non-local iteration domain size? What happens when compute at axis is non-broadcast but local id is?

File: torch/csrc/jit/codegen/cuda/lower_allocation.cpp [hunk 4]
```diff
@@ -195,7 +200,10 @@ class AllocationInserter : public kir::MutableIrVisitor {
           continue;
         }
       }
-      alloc_dims.push_back(ca_id->rawExtent());
+      // TODO: Is this correct? Should allocations be based off of concrete
+      // sizes?
```
Comment (csarofeen) [inline torch/csrc/jit/codegen/cuda/lower_allocation.cpp]: @naoyam same thing here.

File: torch/csrc/jit/codegen/cuda/lower_utils.cpp [hunk 5]
```diff
@@ -168,9 +168,12 @@ ParallelTypeBitmap getParallelBroadcastDomains(
 
 namespace loop_utils {
 
+// TODO: Clean this up, Naoya added a mechanism we should be able to reuse.
 std::pair<kir::ForLoop*, int64_t> getAllocPoint(
```
Comment (csarofeen) [inline torch/csrc/jit/codegen/cuda/lower_utils.cpp:173]: @naoyam could you please double check the logic in here, worried I may be messing up the alloc point like I did with the lower alloc pass.

File: torch/csrc/jit/codegen/cuda/predicate_compute.cpp [hunk 6]
```diff
@@ -113,7 +113,6 @@ kir::Bool* PredicateCompute::getInlinePredicate(
     const ComputeAtRootDomainMap& ca_root_map,
     bool ignore_block_grid_reductions) {
   FUSER_PERF_SCOPE("getInlinePredicate");
-
```
Comment (csarofeen) [inline torch/csrc/jit/codegen/cuda/predicate_compute.cpp:116]: Minor changes should be reverted.

File: torch/csrc/jit/codegen/cuda/lower_allocation.cpp (hunk 4 above)
Comment (csarofeen) [reply torch/csrc/jit/codegen/cuda/lower_allocation.cpp]: Comments posted in reverse, see below, sorry.

File: torch/csrc/jit/codegen/cuda/lower_allocation.cpp (hunk 3 above)
Comment (naoyam) [reply torch/csrc/jit/codegen/cuda/lower_allocation.cpp:121]: It should be safe to ignore broadcast IDs. I suspect that isn't simply taken into consideration in our current implementation.

This may be related: #530

Comment (naoyam) [reply torch/csrc/jit/codegen/cuda/lower_allocation.cpp:121]: Hmm, broadcast *is* ignored in `createAllocExpr`. Maybe there is some bug around here.

Comment (naoyam) [reply torch/csrc/jit/codegen/cuda/lower_allocation.cpp:121]: It seems that we do create an initialization expression with a broadcast loop but don't generate it.

https://github.com/csarofeen/pytorch/blob/20_12_3_devel/torch/csrc/jit/codegen/cuda/codegen.cpp#L593

So, it's no harm, but seems to make more sense to skip such domains at this point.

Comment (naoyam) [reply torch/csrc/jit/codegen/cuda/lower_allocation.cpp:121]: I'll create a small PR to ignore broadcast domains. 

Aside from the case of local-broadcast axes, is there any other concern here?

Comment (naoyam) [reply torch/csrc/jit/codegen/cuda/lower_allocation.cpp:121]: Please see #601

Comment (csarofeen) [reply torch/csrc/jit/codegen/cuda/lower_allocation.cpp:121]: Thank you.

File: test/cpp/jit/test_gpu.cpp [hunk 7]
```diff
@@ -353,9 +354,9 @@ TEST(NVFuserTest, FusionExprEvalPostLower_CUDA) {
   tv0->computeAt(tv3, 1);
   tv1->computeAt(tv3, 1);
 
-  tv3->axis(0)->parallelize(ParallelType::BIDx);
-  tv2->axis(1)->parallelize(ParallelType::Unroll);
-  tv3->axis(1)->parallelize(ParallelType::Unroll);
+  // tv3->axis(0)->parallelize(ParallelType::BIDx);
+  // tv2->axis(1)->parallelize(ParallelType::Unroll);
+  // tv3->axis(1)->parallelize(ParallelType::Unroll);
```
Comment (csarofeen) [inline test/cpp/jit/test_gpu.cpp]: Revert

File: test/cpp/jit/test_gpu.cpp [hunk 8]
```diff
@@ -3956,6 +3957,11 @@ TEST(NVFuserTest, FusionSimpleBCast1_CUDA) {
   tv0->computeAt(tv7, -1);
   tv2->computeAt(tv7, -1);
 
+  std::vector<TensorView*> tvs = {tv0, tv1, tv2, tv3, tv4, tv5, tv6, tv7};
+  for (auto tv : tvs) {
+    // tv->setMemoryType(MemoryType::Global);
+  }
+
```
Comment (csarofeen) [inline test/cpp/jit/test_gpu.cpp]: Revert

File: test/cpp/jit/test_gpu.cpp [hunk 9]
```diff
@@ -4599,6 +4605,9 @@ TEST(NVFuserTest, FusionAdvancedIndexing7_CUDA) {
 
   tv4->axis(0)->parallelize(ParallelType::TIDx);
 
+  fusion.printMath();
+  fusion.printKernel();
```
Comment (csarofeen) [inline test/cpp/jit/test_gpu.cpp]: Revert

File: test/cpp/jit/test_gpu.cpp [hunk 10]
```diff
@@ -4618,6 +4627,7 @@ TEST(NVFuserTest, FusionAdvancedIndexing7_CUDA) {
       &fusion, cg_outputs, {at_t0, at_t1}, {aten_output}, __LINE__, __FILE__);
 }
 
+// TODO: Enable test
```
Comment (csarofeen) [inline test/cpp/jit/test_gpu.cpp]: TODO

File: test/cpp/jit/test_gpu.cpp [hunk 11]
```diff
@@ -4639,30 +4649,84 @@ TEST(NVFuserTest, FusionAdvancedIndexing8_CUDA) {
...
+  // auto aten_output = (at_t0.unsqueeze(-1).expand({numel_x, numel_y}) + at_t1)
+  //                        .to(at::kDouble)
+  //                        .sum();
+
+  // testValidate(
+  //     &fusion, cg_outputs, {at_t0, at_t1}, {aten_output}, __LINE__,
+  //     __FILE__);
+}
+
+// TODO: Enable test
```
Comment (csarofeen) [inline test/cpp/jit/test_gpu.cpp:4767]: TODO

File: torch/csrc/jit/codegen/cuda/executor.cpp [hunk 12]
```diff
@@ -138,6 +138,7 @@ void FusionExecutor::compileFusion(Fusion* fusion, CompileOptions options) {
   }
 
   const auto kernel_code = codegen::generateCudaKernel(kernel, kernelName());
+  // std::cout<<kernel_code<<std::endl;
```
Comment (csarofeen) [inline torch/csrc/jit/codegen/cuda/executor.cpp]: Revert

File: torch/csrc/jit/codegen/cuda/index_compute.cpp [hunk 13]
```diff
@@ -429,8 +452,9 @@ IndexCompute IndexCompute::updateIndexCompute(
 
   const auto gpu_lower = GpuLower::current();
```
Comment (csarofeen) [inline torch/csrc/jit/codegen/cuda/index_compute.cpp:453]: TODO: remove new_index_entries in this function

File: torch/csrc/jit/codegen/cuda/index_compute.cpp [hunk 14]
```diff
@@ -659,16 +699,16 @@ void IndexSwizzle::run() {
...
-        GpuLower::current()->lowerValue(id_to_swizzle_j)->as<kir::IterDomain>();
+        gpu_lower->lowerValue(id_to_swizzle_j)->as<kir::IterDomain>();
 
     if (indexMap().find(id_to_swizzle_i_kir) != indexMap().end() &&
         indexMap().find(id_to_swizzle_j_kir) != indexMap().end()) {
       auto idx_to_swizzle_i = indexMap().at(id_to_swizzle_i_kir);
       auto idx_to_swizzle_j = indexMap().at(id_to_swizzle_j_kir);
 
-      kir::IrBuilder ir_builder(GpuLower::current()->kernel());
+      kir::IrBuilder ir_builder(gpu_lower->kernel());
```
Comment (csarofeen) [inline torch/csrc/jit/codegen/cuda/index_compute.cpp]: Move up with gpu_lower definition for consistency

File: torch/csrc/jit/codegen/cuda/index_compute.cpp [hunk 15]
```diff
@@ -695,410 +735,61 @@ void IndexSwizzle::handle(Expr* e) {
...
+  auto pairwiseMap = PairwiseRootDomainMap(producer_tv, consumer_tv);
+  auto producerAsC =
+      TransformReplay::replayPasC(
+          producer_tv->domain(), consumer_tv->domain(), -1, pairwiseMap)
+          .first;
 
-  // grab all tensor views from producer_tv <- computeAtRoot
-  auto tv_stack = getComputeAtTVStackFrom(consumer_tv);
-  tv_stack.push_back(producer_tv);
+  // Cannot be done after domain_guard
```
Comment (csarofeen) [inline torch/csrc/jit/codegen/cuda/index_compute.cpp]: TODO: Check this, domain guard shouldn't impact root mapping.

File: torch/csrc/jit/codegen/cuda/index_compute.cpp [hunk 16]
```diff
@@ -1209,28 +902,112 @@ std::unordered_map<kir::ForLoop*, kir::Val*> indexMapFromTV(
...
-      indexMapFromTV(producer_tv, loops);
-
-  auto index_and_extent_map = generateIndexAndExtentMap(
-      tv_stack,
-      std::deque<kir::ForLoop*>(loops.begin(), loops.end()),
-      loop_to_ind_map,
-      std::vector<bool>(producer_tv->getRootDomain().size(), false),
-      ca_root_map,
-      true,
+  // Cannot be done after domain_guard
```
Comment (csarofeen) [inline torch/csrc/jit/codegen/cuda/index_compute.cpp]: TODO: Also check.

File: torch/csrc/jit/codegen/cuda/index_compute.cpp [hunk 17]
```diff
@@ -1384,33 +1179,139 @@ kir::TensorIndex* Index::getGlobalConsumerIndex(
...
+  // This is a special case with consumer tensor views that are a result of an
+  // rfactor. If we only use the reference tensor doesn't necessarily know the
+  // difference between reduction axes and iteration axes. So if it propagates
+  // zero merged in indices, it doesn't know not to propagate them from
+  // reduction IDs towards iter IDs. Since producer tensors won't have indices
+  // associated with their reduction axes (reductions have to be finished before
+  // they can be consumed) and since global consumers won't have zero merge in
+  // propagation this can only happen here. We will go through and zero out any
+  // for loops that are a reduction axis for the consumer, first find them.
+  {
```
Comment (csarofeen) [inline torch/csrc/jit/codegen/cuda/index_compute.cpp]: TODO: compare with producer for correctness

File: torch/csrc/jit/codegen/cuda/index_compute.cpp [hunk 18]
```diff
@@ -1384,33 +1179,139 @@ kir::TensorIndex* Index::getGlobalConsumerIndex(
...
+  for (size_t loop_i; loop_i < loops.size(); loop_i++) {
+    auto ref_axis = gpu_lower->lowerValue(reference_domain->axis(loop_i))
+                        ->as<kir::IterDomain>();
+    ref_id_to_ind_map[ref_axis] = loop_to_ind_map[loops[loop_i]];
+  }
+
+  // Map reference tensor to consumer
+  std::unordered_map<IterDomain*, IterDomain*> root_ref_to_consumer;
+  for (auto c_root : consumer_tv->getMaybeRFactorDomain()) {
+    auto concrete_id = gpu_lower->caIndexMap().getConcreteMappedID(c_root);
```
Comment (csarofeen) [inline torch/csrc/jit/codegen/cuda/index_compute.cpp:1218]: TODO: Is this okay for consumer, though not for producer?

File: torch/csrc/jit/codegen/cuda/index_compute.cpp [hunk 19]
```diff
@@ -1384,33 +1179,139 @@ kir::TensorIndex* Index::getGlobalConsumerIndex(
...
+  // }
+
+  // Index into consumer using reference indexing
+  auto consumer_indexing = ref_compute.updateIndexCompute(
+      consumer_tv->domain(),
+      ref_2_consumer,
+      {},
+      consumer_tv->domain()->contiguity());
+
+  // TODO: Re-enable swizzle
```
Comment (csarofeen) [inline torch/csrc/jit/codegen/cuda/index_compute.cpp]: TODO: Swizzle.

File: torch/csrc/jit/codegen/cuda/lower2device.cpp [hunk 20]
```diff
@@ -126,21 +148,21 @@ void GpuLower::lower() {
...
   // Reorder expressions for loop-nest generation respecting computeAt
   // relationships
-  const auto reordered_exprs = reorderExprsForComputeAt(fusion_->exprs());
+  auto sorted_exprs = reorderExprsTest();
 
   // Generate loop-nests and place each expression at its
   // corresponding loop
-  const auto lowered_exprs =
-      LoopNestGenerator::loweredExprs(fusion_, reordered_exprs);
-
```
Comment (csarofeen) [inline torch/csrc/jit/codegen/cuda/lower2device.cpp:135]: revert const

File: torch/csrc/jit/codegen/cuda/lower2device.h [hunk 21]
```diff
@@ -36,6 +38,22 @@ class TORCH_CUDA_API GpuLower {
   //! (or nullptr if no lowering is in progress)
   static GpuLower* current();
 
+  const ComputeAtRootDomainMap& caRootMap() const {
```
Comment (csarofeen) [inline torch/csrc/jit/codegen/cuda/lower2device.h]: Do we still use this anywhere? Would be good if it was non ca version.

File: torch/csrc/jit/codegen/cuda/lower_allocation.cpp [hunk 22]
```diff
@@ -113,10 +112,12 @@ class AllocationInserter : public kir::MutableIrVisitor {
...
-      auto ca_id =
-          gpu_lower->lowerValue(fuser_tv->getComputeAtAxis(axis_i).first)
+      auto concrete_id =
+          gpu_lower
+              ->lowerValue(gpu_lower->caParallelMap().getConcreteMappedID(
+                  fuser_tv->axis(axis_i)))
               ->as<kir::IterDomain>();
-      init_dims.push_back(ca_id);
+      init_dims.push_back(concrete_id);
     }
```
Comment (csarofeen) [inline torch/csrc/jit/codegen/cuda/lower_allocation.cpp:122]: TODO: make sure this comes through https://github.com/csarofeen/pytorch/pull/601

File: torch/csrc/jit/codegen/cuda/lower_allocation.cpp (hunk 4 above)
Comment (csarofeen) [reply torch/csrc/jit/codegen/cuda/lower_allocation.cpp]: I think this should probably be indexMap concrete, not loop map

Comment (csarofeen) [reply torch/csrc/jit/codegen/cuda/lower_allocation.cpp]: TODO: change, run memcheck

File: torch/csrc/jit/codegen/cuda/lower_compute_at_map.cpp [hunk 23]
```diff
@@ -0,0 +1,658 @@
...
+    return id->parallelType();
+  }
+  auto parallel_type_it = kir_parallel_type_map_.find(disjoint_set_it->second);
+  if (parallel_type_it == kir_parallel_type_map_.end()) {
+    return id->parallelType();
+  }
+  return parallel_type_it->second;
+}
+
+// std::unordered_map<IterDomain*, IterDomain*> ComputeAtMap::mapFromTo(
```
Comment (csarofeen) [inline torch/csrc/jit/codegen/cuda/lower_compute_at_map.cpp]: TODO: Cleanup

File: torch/csrc/jit/codegen/cuda/lower_loops.cpp [hunk 24]
```diff
@@ -221,6 +224,185 @@ void LoopNestGenerator::generate(const std::vector<Expr*>& exprs) {
     handle(expr);
   }
 }
+// ======================================================
```
Comment (csarofeen) [inline torch/csrc/jit/codegen/cuda/lower_loops.cpp]: TODO: Clean up.

File: test/cpp/jit/test_gpu.cpp [hunk 25]
```diff
@@ -1,4 +1,4 @@
-#if defined(USE_CUDA)
+// #if defined(USE_CUDA)
```
Comment (csarofeen) [inline test/cpp/jit/test_gpu.cpp]: Revert

File: test/cpp/jit/test_gpu.cpp [hunk 26]
```diff
@@ -11424,7 +11491,64 @@ TEST(NVFuserTest, FusionTransposeWithSwizzle1DThreadBlock_CUDA) {
...
+  //       __FILE__,
+  //       "",
+  //       lparams);
+}
+
 } // namespace jit
 } // namespace torch
 
-#endif // #if defined(USE_CUDA)
+// #endif // #if defined(USE_CUDA)
```
Comment (csarofeen) [inline test/cpp/jit/test_gpu.cpp]: Revert

File: tools/build_variables.bzl [hunk 27]
```diff
@@ -371,6 +371,7 @@ libtorch_cuda_sources = [
     "torch/csrc/jit/codegen/cuda/fusion.cpp",
     "torch/csrc/jit/codegen/cuda/graph_fuser.cpp",
     "torch/csrc/jit/codegen/cuda/index_compute.cpp",
+    "torch/csrc/jit/codegen/cuda/test_index_compute.cpp",
```
Comment (csarofeen) [inline tools/build_variables.bzl]: TODO: Rename

File: tools/build_variables.bzl [hunk 28]
```diff
@@ -405,6 +407,7 @@ libtorch_cuda_sources = [
     "torch/csrc/jit/codegen/cuda/register_interface.cpp",
     "torch/csrc/jit/codegen/cuda/root_domain_map.cpp",
     "torch/csrc/jit/codegen/cuda/scheduler.cpp",
+    "torch/csrc/jit/codegen/cuda/segmenter_helper.cpp",
```
Comment (csarofeen) [inline tools/build_variables.bzl]: TODO: Rename

File: test/cpp/jit/test_gpu.cpp (hunk 11 above)
Comment (csarofeen) [reply test/cpp/jit/test_gpu.cpp:4767]: Requires changes to root mapping. CC @naoyam . Probably should do this in a follow up.

File: torch/csrc/jit/codegen/cuda/kernel_ir_printer.h [hunk 29]
```diff
@@ -21,12 +21,15 @@ namespace kir {
 //! This class is intended for debug printing, so it attempts
```
Comment (csarofeen) [inline torch/csrc/jit/codegen/cuda/kernel_ir_printer.h]: TODO: Revert changes, or split out to a separate PR.

File: torch/csrc/jit/codegen/cuda/lower2device.cpp [hunk 30]
```diff
@@ -109,14 +109,30 @@ void GpuLower::lower() {
   // Compute thread predicates
   ThreadPredicateMap preds(fusion_);
 
-  // Compute root-domain mappings
-  ComputeAtRootDomainMap ca_root_map;
-  ca_root_map.build();
+  // In the future we may directly use this map, but for now it will propagate
+  // and validate (to some extent) the parallelization strategy.
+  // TODO: Does this have to go before the next mapping?
```
Comment (csarofeen) [inline torch/csrc/jit/codegen/cuda/lower2device.cpp]: Fix comment, no longer TODO, has to be done first since right now we're setting parallel strategy on fusion iter domains and need to do it before lowering.

File: torch/csrc/jit/codegen/cuda/lower_allocation.cpp (hunk 4 above)
Comment (csarofeen) [reply torch/csrc/jit/codegen/cuda/lower_allocation.cpp]: Remove TODO, concrete size in index map cannot be larger than local size.

File: torch/csrc/jit/codegen/cuda/lower_loops.cpp [hunk 31]
```diff
@@ -17,17 +18,22 @@ namespace jit {
...
+
+LoopNestGenerator::LoopNestGenerator(const std::vector<Expr*>& exprs) {
   generate(exprs);
 }
 
 namespace {
 
 // TODO(kir): revisit and try to simplify this
-kir::ForLoop* openForHelper(kir::ForLoop* scope, IterDomain* id) {
+kir::ForLoop* openForHelper2(kir::ForLoop* scope, IterDomain* id) {
```
Comment (csarofeen) [inline torch/csrc/jit/codegen/cuda/lower_loops.cpp]: TODO: Fix name back.

File: torch/csrc/jit/codegen/cuda/lower_utils.cpp (hunk 5 above)
Comment (naoyam) [reply torch/csrc/jit/codegen/cuda/lower_utils.cpp:173]: This looks good to me.

File: torch/csrc/jit/codegen/cuda/lower_utils.cpp [hunk 32]
```diff
@@ -168,9 +168,12 @@ ParallelTypeBitmap getParallelBroadcastDomains(
 
 namespace loop_utils {
 
+// TODO: Clean this up, Naoya added a mechanism we should be able to reuse.
```
Comment (naoyam) [inline torch/csrc/jit/codegen/cuda/lower_utils.cpp:172]: Which mechanism are you referring to?

File: torch/csrc/jit/codegen/cuda/lower_loops.cpp [hunk 33]
```diff
@@ -88,137 +103,83 @@ void LoopNestGenerator::handle(const Expr* expr) {
...
-    // doesn't
-    loop_structure.push_back(out->getComputeAtAxis(out_i).first);
+    loop_structure.push_back(out_tv->axis((int)out_i));
   }
 
-  // At this point loop_structure contains our overal target loop nest structure
-  // Lets get a copy of the loop structure, and figure out which loops we need
-  // to open.
-  auto loops_to_open = loop_structure;
+  // Save position of out_id_it as we will append to loop structure
```
Comment (naoyam) [inline torch/csrc/jit/codegen/cuda/lower_loops.cpp]: Shouldn't this be placed before the above for loop (i.e., before line 148)?

File: torch/csrc/jit/codegen/cuda/lower_loops.cpp [hunk 34]
```diff
@@ -88,137 +103,83 @@ void LoopNestGenerator::handle(const Expr* expr) {
...
-      for (start = 0; start < ca_view->nDims(); start++) {
-        if (loop_structure.back() == ca_view->getComputeAtAxis(start).first) {
-          break;
-        }
-      }
-      start++;
-    }
+    // Safe to use loop map since this is outside the compute at point
+    auto concrete_id =
+        gpu_lower->caParallelMap().getConcreteMappedID(out_tv->axis(out_i));
```
Comment (naoyam) [inline torch/csrc/jit/codegen/cuda/lower_loops.cpp:126]: Shouldn't this be `caLoopMap()` instead of `caParallelMap()`?

File: torch/csrc/jit/codegen/cuda/lower_utils.cpp (hunk 32 above)
Comment (csarofeen) [reply torch/csrc/jit/codegen/cuda/lower_utils.cpp:172]: I thought we may want to reuse `lower_allocation.cpp::findAllocationPosition`

Comment (naoyam) [reply torch/csrc/jit/codegen/cuda/lower_utils.cpp:172]: I'll try to refactor them.

File: torch/csrc/jit/codegen/cuda/lower_compute_at_map.cpp [hunk 35]
```diff
@@ -0,0 +1,578 @@
...
+          auto p_id = entry.second;
+          // If outside CA point and we're creating parallel map, do not map the
+          // axis
+          if (mapping_mode_ == MappingMode::PARALLEL &&
+              right_of_ca_point.find(p_id) != right_of_ca_point.end()) {
+            continue;
+          }
+          // Map the id's together
+          map_ids(p_id, c_id);
+          p2c_map[p_id] = c_id;
```
Comment (naoyam) [inline torch/csrc/jit/codegen/cuda/lower_compute_at_map.cpp]: c2p_map is not one-to-one when the mapping mode is LOOP, i.e., when bcast mismatch is allowed. Doesn't it cause p2c_map non-deterministic? There can be multiple c_id for a single p_id. The final mapping for the p_id would depend on the internal ordering of c2p_map.

File: torch/csrc/jit/codegen/cuda/lower_expr_sort.h [hunk 36]
```diff
@@ -7,9 +7,9 @@ namespace jit {
 namespace fuser {
 namespace cuda {
 
-std::vector<Expr*> reorderExprsForComputeAt(const std::vector<Expr*>& exprs);
+TORCH_CUDA_API std::vector<Expr*> reorderExprsTest();
```
Comment (naoyam) [inline torch/csrc/jit/codegen/cuda/lower_expr_sort.h]: Is there any reason to name this function `-Test`?

File: torch/csrc/jit/codegen/cuda/lower_compute_at_map.cpp (hunk 35 above)
Comment (csarofeen) [reply torch/csrc/jit/codegen/cuda/lower_compute_at_map.cpp]: There are multiple p_id's it could map to, but the mapping will simply forward from best effort replay. So the most bcast merged axis in producer should be mapped to the corresponding c_id.
I'm not 100%, but I'm pretty sure it should be deterministic.

File: torch/csrc/jit/codegen/cuda/lower_loops.cpp (hunk 34 above)
Comment (csarofeen) [reply torch/csrc/jit/codegen/cuda/lower_loops.cpp:126]: I actually think it should be index or parallel. Within the compute at they should have the maximum extent that would be needed.

Comment (csarofeen) [reply torch/csrc/jit/codegen/cuda/lower_loops.cpp:126]: No, you're right, it should be loop because of the misaligned bcast behavior.

File: torch/csrc/jit/codegen/cuda/lower_loops.cpp (hunk 33 above)
Comment (csarofeen) [reply torch/csrc/jit/codegen/cuda/lower_loops.cpp]: Yes, it should be.

File: torch/csrc/jit/codegen/cuda/lower_expr_sort.h (hunk 36 above)
Comment (csarofeen) [reply torch/csrc/jit/codegen/cuda/lower_expr_sort.h]: No of course not, just forgot to rename it.

File: torch/csrc/jit/codegen/cuda/lower_compute_at_map.cpp (hunk 35 above)
Comment (naoyam) [reply torch/csrc/jit/codegen/cuda/lower_compute_at_map.cpp]: Apparently, there are multiple different c_ids for a single p_id. I inserted this:

```
diff --git a/torch/csrc/jit/codegen/cuda/lower_compute_at_map.cpp b/torch/csrc/jit/codegen/cuda/lower_compute_at_map.cpp
index 6001e02dc5..c320ee9781 100644
--- a/torch/csrc/jit/codegen/cuda/lower_compute_at_map.cpp
+++ b/torch/csrc/jit/codegen/cuda/lower_compute_at_map.cpp
@@ -287,6 +287,7 @@ void ComputeAtMap::build() {
           }
           // Map the id's together
           mapIds(p_id, c_id);
+          TORCH_INTERNAL_ASSERT(p2c_map.find(p_id) == p2c_map.end() || p2c_map.at(p_id) == c_id);
           p2c_map[p_id] = c_id;
         }
```

And I got assertion failures with it. So, the mapping of the p_id would just depend on the ordering among the multiple c_ids, which would be non-deterministic.

File: torch/csrc/jit/codegen/cuda/lower_loops.cpp (hunk 34 above)
Comment (naoyam) [reply torch/csrc/jit/codegen/cuda/lower_loops.cpp:126]: Hmm, I tried `caLoopMap()` and got a validation error with `FusionPersistentNormLocalShared`. Will look into it.

File: torch/csrc/jit/codegen/cuda/lower_expr_sort.cpp [hunk 37]
```diff
@@ -1,372 +1,904 @@
...
+ExprGroup* ExprSegmentationSorter::makeMergedNode(
+    ExprGroup* sg1,
+    ExprGroup* sg2) {
+  std::vector<IterDomain*> resulting_ca_axes;
+  auto& domain1 = sg1->payload()->ca_domains;
+  auto& domain2 = sg2->payload()->ca_domains;
+  auto it1 = domain1.begin();
+  auto it2 = domain2.begin();
+
+  while (it1 != domain1.end() && it2 != domain2.end()) {
```
Comment (naoyam) [inline torch/csrc/jit/codegen/cuda/lower_expr_sort.cpp]: Shouldn't this use `||` instead of `&&`? With `&&`, the first two if clauses are never used.

Comment (naoyam) [reply torch/csrc/jit/codegen/cuda/lower_expr_sort.cpp]: Can we just do something like:

```
auto resulting_ca_axes = domain1;
for (auto id2 : domain2) {
  if (std::none_of(domain1.begin(), domain1.end(), [&](auto id1) {
         return caIndexMap().areMapped(id1, id2);
      })) {
      resulting_ca_axes.push_back(id2);
  }
}
```

Comment (csarofeen) [reply torch/csrc/jit/codegen/cuda/lower_expr_sort.cpp]: We need all iteration domains from both to propagate, and we need to maintain relative order while trying to reduce duplicates when we can. I added some more comments.

File: torch/csrc/jit/codegen/cuda/lower_expr_sort.h (hunk 36 above)
Comment (csarofeen) [reply torch/csrc/jit/codegen/cuda/lower_expr_sort.h]: Reverted name.

File: torch/csrc/jit/codegen/cuda/lower_loops.cpp (hunk 33 above)
Comment (csarofeen) [reply torch/csrc/jit/codegen/cuda/lower_loops.cpp]: Moved.

File: torch/csrc/jit/codegen/cuda/lower_expr_sort.cpp (hunk 37 above)
Comment (csarofeen) [reply torch/csrc/jit/codegen/cuda/lower_expr_sort.cpp]: You were right about the `||` instead of `&&`, changed that as well.

File: torch/csrc/jit/codegen/cuda/lower_compute_at_map.cpp (hunk 35 above)
Comment (csarofeen) [reply torch/csrc/jit/codegen/cuda/lower_compute_at_map.cpp]: Fixed this with https://github.com/csarofeen/pytorch/pull/597/commits/6f279a2ecb5a135b4ef9d99fbfacb75af60c8673

File: torch/csrc/jit/codegen/cuda/lower_expr_sort.cpp [hunk 38]
```diff
@@ -1,372 +1,963 @@
...
+ private:
+  // "Ancestor nodes", towards inputs of segmentedDAG
+  std::vector<ExprGroupConnections*> producer_edges_;
+
+  // "Descendent nodes", towards outputs of segmentedDAG
+  std::vector<ExprGroupConnections*> consumer_edges_;
+
+  // TODO: Used?
+  std::vector<Val*> input_vals_;
+  std::vector<Val*> output_vals_;
```
Comment (naoyam) [inline torch/csrc/jit/codegen/cuda/lower_expr_sort.cpp]: Are these used? They are set, but they do not seem to be used. Can they be removed?

Comment (csarofeen) [reply torch/csrc/jit/codegen/cuda/lower_expr_sort.cpp]: If they're not used, then yes, please remove.

File: torch/csrc/jit/codegen/cuda/lower_loops.cpp (hunk 34 above)
Comment (csarofeen) [reply torch/csrc/jit/codegen/cuda/lower_loops.cpp:126]: Parallel map should be fine for now. Although in the future we'll want to move to generating a reference either before or during the loop construction and including the reference iter domains in the iter domain maps. The issue that could happen from here is if one of the TensorViews doesn't resolve all the broadcasts that are resolved in a loop nest. This doesn't happen in our current testing but could be done.

File: torch/csrc/jit/codegen/cuda/lower_expr_sort.cpp (hunk 38 above)
Comment (naoyam) [reply torch/csrc/jit/codegen/cuda/lower_expr_sort.cpp]: Removed.

File: torch/csrc/jit/codegen/cuda/index_compute.cpp [hunk 39]
```diff
@@ -444,6 +465,10 @@ IndexCompute IndexCompute::updateIndexCompute(
       updated_index_map[new_id] = index_map_.at(prev_id);
     }
 
+    if (!prev_id->isBroadcast() && new_id->isBroadcast()) {
```
Comment (naoyam) [inline torch/csrc/jit/codegen/cuda/index_compute.cpp:469]: Can you give some comments here? `updated_extent_map` is set at three different paths with `extent_map_.at(prev_id)` or `getExtent(prev_id)`. Would be nice why each path would be taken and why it would use `extent_map_` or `getExtent`.

Comment (csarofeen) [reply torch/csrc/jit/codegen/cuda/index_compute.cpp:469]: I think this should actually be an `if else` in the `if` below. This is primarily just not setting an entry in the extent map unless it's needed. That way we know if extent has diverged from what's in the `IterDomain`. `getExtent` is just a convenience wrapper that will return a map entry if there is one, otherwise will return the extent of the `IterDomain` provided.

Comment (csarofeen) [reply torch/csrc/jit/codegen/cuda/index_compute.cpp:469]: The one I actually don't get below is the is reduction section, uncertain if or why that'd be needed. Another thing to keep in mind, is I think we only update the index compute once per indexing now, instead of through out a large tv stack so we may be able to simplify this code somewhat.

Comment (naoyam) [reply torch/csrc/jit/codegen/cuda/index_compute.cpp:469]: Thanks. I think this part of the change is also related. The propagation through the `isReduction()` logic may not kick in now, and that seems to explain why an axis can be missing in `extent_map` .

Comment (naoyam) [reply torch/csrc/jit/codegen/cuda/index_compute.cpp:469]: I'll revisit this later again.

**Comment by naoyam** [comment]
I'm mostly done reviewing this PR. There are mainly three things to do, all of which can be done as separate PRs.

- Enable the AdvancedLowering3 test. Root mapping seems to have an issue. https://github.com/csarofeen/pytorch/pull/597#discussion_r560349607
- Refactoring of allocation position finders. https://github.com/csarofeen/pytorch/pull/597#discussion_r562038955
- IndexCompute::updateIndexCompute. https://github.com/csarofeen/pytorch/pull/597#discussion_r566519607

We may also revisit this: https://github.com/csarofeen/pytorch/pull/597#discussion_r563782875.

**Review by naoyam — APPROVED** [review]
